fw/ringbuffer.c

```diff
@@ -0,0 +1,248 @@
+/**
+ *		Tempesta FW
+ *
+ * Handling ring buffers mapped to userspace.
+ *
+ * Copyright (C) 2024 Tempesta Technologies, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License,
+ * or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 59
+ * Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+ */
+
+#include "ringbuffer.h"
+#include "lib/str.h"
+#include <linux/types.h>
+#include <asm/io.h>
+#include <linux/log2.h>
+#include <linux/cpumask.h>
+
+typedef struct {
+	TfwRingbuffer *rb;
+	struct page *pg;
+	atomic_t unmapped;
+} TfwRingbufferArea;
+
+#define AREA_SIZE(size) (size + TFW_RINGBUFFER_DATA_OFFSET)
+#define RBA_AREA_SIZE(rba) AREA_SIZE((rba)->rb->size)
+
+static DEFINE_PER_CPU(TfwRingbufferArea, rb_area);
+static bool proc_file_is_open;
+
+int
+tfw_ringbuffer_write(TfwStr **strs, unsigned int count)
+{
+	TfwRingbufferArea *rba = this_cpu_ptr(&rb_area);
+	TfwRingbuffer *rb = rba->rb;
+	unsigned int i;
+	u64 head, tail;
+	u32 full_size = 0;
+
+	if (atomic_read(&rba->unmapped))
+		return -EAGAIN;
+
+	preempt_disable();
+
+	head = rb->head;
+	tail = smp_load_acquire(&rb->tail);
+
+	for (i = 0; i < count; ++i)
+		full_size += strs[i]->len;
+
+	if (head - tail + full_size >= rb->size) {
+		preempt_enable();
+		return -ENOMEM;
+	}
+
+	for (i = 0; i < count; ++i) {
+		TfwStr *c, *end;
+
+		TFW_STR_FOR_EACH_CHUNK(c, strs[i], end) {
+			char *data = c->data;
+			u32 len = c->len;
+
+			while (len) {
+				u32 mhead, cur_len;
+
+				mhead = head & rb->mask;
+				cur_len = min(len, rb->size - mhead);
+
+				memcpy_fast(rb->data + mhead, data, cur_len);
+				data += cur_len;
+				head += cur_len;
+				len -= cur_len;
+			}
+		}
+	}
+
+	smp_store_release(&rb->head, head);
+	smp_mb();
```

Comment:
Why do you use mem barrier with per-cpu data?

Reply:
You are right, it always executes in softirq context, and preemption disabling is excessive. Will get rid of it.

Reply:
I think the point is that atomics imply full barriers on x86-64

Reply:
Oops, I answered to the comment about preempt_disable.

So we can do without smp_mb, don't we?

Reply:
> So we can do without smp_mb, don't we?

Yes, we don't need `smp_mb`